code/modules/mob/living/basic/icemoon/ice_demon/ice_demon_ai.dm

```diff
@@ -0,0 +1,153 @@
+/datum/ai_controller/basic_controller/ice_demon
+	blackboard = list(
+		BB_TARGETTING_DATUM = new /datum/targetting_datum/basic,
+		BB_LIST_SCARY_ITEMS = list(
+			/obj/item/weldingtool,
+			/obj/item/flashlight/flare,
+		),
+		BB_BASIC_MOB_FLEEING = TRUE,
+		BB_MINIMUM_DISTANCE_RANGE = 3,
+	)
+
+	ai_movement = /datum/ai_movement/basic_avoidance
+	idle_behavior = /datum/idle_behavior/idle_random_walk
+	planning_subtrees = list(
+		/datum/ai_planning_subtree/simple_find_target,
+		/datum/ai_planning_subtree/flee_target/ice_demon,
+		/datum/ai_planning_subtree/teleport_away_from_target,
+		/datum/ai_planning_subtree/maintain_distance/cover_minimum_distance/ice_demon,
+		/datum/ai_planning_subtree/ranged_skirmish,
+		/datum/ai_planning_subtree/find_and_hunt_target/teleport_destination,
+		/datum/ai_planning_subtree/targeted_mob_ability/summon_afterimages,
+	)
+
+
+/datum/ai_planning_subtree/maintain_distance/cover_minimum_distance/ice_demon
+	maximum_distance = 7
+
+
+/datum/ai_planning_subtree/teleport_away_from_target/SelectBehaviors(datum/ai_controller/controller, seconds_per_tick)
+	if(!controller.blackboard_key_exists(BB_BASIC_MOB_CURRENT_TARGET))
+		return
+	var/atom/target = controller.blackboard[BB_BASIC_MOB_CURRENT_TARGET]
+	var/distance_from_target = get_dist(target, controller.pawn)
+	if(distance_from_target >= controller.blackboard[BB_MINIMUM_DISTANCE_RANGE])
+		controller.clear_blackboard_key(BB_ESCAPE_DESTINATION)
+		return
+	var/datum/action/cooldown/ability = controller.blackboard[BB_DEMON_TELEPORT_ABILITY]
```

Comment:
Any reason this couldn't be a generic behaviour rather than in this file, with a blackboard key for some kind of movement ability as a var?